dhis2/dhis2-core
# feat: job progress tracking for data validation job [DHIS2-12858]

### Summary
Adds `JobProgress` tracking to the data validation job (a.k.a. "monitoring" job).

Because of the already present parallel processing that needed to be replaced with the progress own parallel processing to be able to track this a more extensive change was required. This cascaded a bit as it then required reorganization to get correct `@Transactional` boundaries. In the process doing that I cleaned up to understand the code I was changing. 

### Manual Testing
* goto scheduler app
* create new job of type _Monitoring_ (I picked _Validation rule groups_ `ANC` and start/end `0`)
* run the job manually
* check the running status http://localhost:8080/api/scheduling/running/MONITORING
* check the completed status http://localhost:8080/api/scheduling/completed/MONITORING

### Automatic Testing
Existing tests should cover the validation itself sufficiently. 

### Example Output
```json
[
  {
    "summary":"Data validation done",
    "status":"SUCCESS",
    "completedTime":"2022-03-23T11:18:17.994",
    "startedTime":"2022-03-23T11:18:16.861",
    "description":"Data validation",
    "stages":[
      {
        "summary":"ANC 3 <= ANC 2, ANC 2 <= ANC 1",
        "status":"SUCCESS",
        "completedTime":"2022-03-23T11:18:16.917",
        "startedTime":"2022-03-23T11:18:16.892",
        "description":"Preparing analysis parameters",
        "totalItems":0,
        "items":[
          
        ],
        "duration":25,
        "complete":true
      },
      {
        "status":"SUCCESS",
        "completedTime":"2022-03-23T11:18:17.972",
        "startedTime":"2022-03-23T11:18:17.283",
        "description":"Evaluating validation rules in chunks of 500",
        "totalItems":3,
        "items":[
          {
            "status":"SUCCESS",
            "completedTime":"2022-03-23T11:18:17.961",
            "startedTime":"2022-03-23T11:18:17.284",
            "description":"500-999",
            "duration":677,
            "complete":true
          },
          {
            "status":"SUCCESS",
            "completedTime":"2022-03-23T11:18:17.972",
            "startedTime":"2022-03-23T11:18:17.285",
            "description":"0-499",
            "duration":687,
            "complete":true
          },
          {
            "status":"SUCCESS",
            "completedTime":"2022-03-23T11:18:17.620",
            "startedTime":"2022-03-23T11:18:17.287",
            "description":"1000-1331",
            "duration":333,
            "complete":true
          }
        ],
        "duration":689,
        "complete":true
      },
      {
        "status":"SUCCESS",
        "completedTime":"2022-03-23T11:18:17.988",
        "startedTime":"2022-03-23T11:18:17.972",
        "description":"Reloading attribute option combos",
        "totalItems":0,
        "items":[
          
        ],
        "duration":16,
        "complete":true
      },
      {
        "status":"SUCCESS",
        "completedTime":"2022-03-23T11:18:17.993",
        "startedTime":"2022-03-23T11:18:17.988",
        "description":"Persisting Results",
        "totalItems":0,
        "items":[
          
        ],
        "duration":5,
        "complete":true
      }
    ],
    "jobId":"IzrirTv1i0t",
    "duration":1133,
    "complete":true
  }
]
```
